UM/Qt/QtRenderer.py

```diff
@@ -0,0 +1,192 @@
+import copy
+from ctypes import c_void_p
+
+vertexBufferProperty = "__qtgl2_vertex_buffer"
+indexBufferProperty = "__qtgl2_index_buffer"
+
+##  A Renderer implementation using PyQt's OpenGL implementation to render.
+class QtRenderer(Renderer):
+    def __init__(self):
+        super().__init__()
+
+        self._controller = Application.getInstance().getController()
+        self._scene = self._controller.getScene()
+
+        self._vertex_buffer_cache = {}
+        self._index_buffer_cache = {}
+
+        self._initialized = False
+
+        self._light_position = Vector(0, 0, 0)
+        self._background_color = QColor(128, 128, 128)
+        self._viewport_width = 0
+        self._viewport_height = 0
+        self._window_width = 0
+        self._window_height = 0
+
+        self._batches = []
+
+        self._quad_buffer = None
+
+        self._camera = None
+
+    ##  Get an integer multiplier that can be used to correct for screen DPI.
+    def getPixelMultiplier(self):
+        # Standard assumption for screen pixel density is 96 DPI. We use that as baseline to get
+        # a multiplication factor we can use for screens > 96 DPI.
+        return round(Application.getInstance().primaryScreen().physicalDotsPerInch() / 96.0)
+
+    ##  Get the list of render batches.
+    def getBatches(self):
+        return self._batches
+
+    ##  Overridden from Renderer.
+    #
+    #   This makes sure the added render pass has the right size.
+    def addRenderPass(self, render_pass):
+        super().addRenderPass(render_pass)
+        render_pass.setSize(self._viewport_width, self._viewport_height)
+
+    ##  Set background color used when rendering.
+    def setBackgroundColor(self, color):
+        self._background_color = color
+
+    ##  Set the viewport size.
+    #
+    #   \param width The new width of the viewport.
+    #   \param height The new height of the viewport.
+    def setViewportSize(self, width, height):
+        self._viewport_width = width
+        self._viewport_height = height
+
+        for render_pass in self._render_passes:
+            render_pass.setSize(width, height)
+
+    ##  Set the window size.
+    def setWindowSize(self, width, height):
+        self._window_width = width
+        self._window_height = height
+
+    ##  Get the window size.
+    #
+    #   \return A tuple of (window_width, window_height)
+    def getWindowSize(self):
+        return (self._window_width, self._window_height)
+
+    def beginRendering(self):
+        if not self._initialized:
+            self._initialize()
+
+        self._gl.glViewport(0, 0, self._viewport_width, self._viewport_height)
+        self._gl.glClearColor(self._background_color.redF(), self._background_color.greenF(), self._background_color.blueF(), self._background_color.alphaF())
+        self._gl.glClear(self._gl.GL_COLOR_BUFFER_BIT | self._gl.GL_DEPTH_BUFFER_BIT)
+        self._gl.glClearColor(0.0, 0.0, 0.0, 0.0)
+
+    ##  Put a node in the render queue
```

Comment:
This now forwards to RenderBatch::__init__(). I have updated the docs to be more explicit about that.